lenskit-core/src/main/java/org/grouplens/lenskit/data/dao/packed/BinaryRatingDAO.java

```diff
@@ -138,14 +145,39 @@ public static BinaryRatingDAO open(File file) throws IOException {
             BinaryIndexTable utbl = BinaryIndexTable.fromBuffer(header.getUserCount(), tableBuffer);
             BinaryIndexTable itbl = BinaryIndexTable.fromBuffer(header.getItemCount(), tableBuffer);
 
-            return new BinaryRatingDAO(file, header, data, utbl, itbl);
+            return new BinaryRatingDAO(file, header, data, utbl, itbl,header.getRatingCount(),Long.MAX_VALUE);
         } finally {
             input.close();
         }
     }
+    public  BinaryRatingDAO createWindowedView(long timestamp) {
+
+        if (timestamp >= limitTimestamp) {
+            return this;
+        }
+
+        Rating r = new RatingBuilder().setUserId(0).setItemId(0).setRating(0).setTimestamp(timestamp).build();
+
+        int idx = Collections.binarySearch(getRatingList(), r, Events.TIMESTAMP_COMPARATOR);
+
+        if (idx < 0) {
+            idx = -idx - 1;
+        } else {
+            while (getRatingList().get(idx).getTimestamp() >= timestamp) {
```

Comment:
Calling `getRatingList()` that many times creates a bunch of extra objects. Call it near the beginning of this method.